bpf/http_sock.c

```diff
@@ -241,165 +241,117 @@ int BPF_KPROBE(kprobe_sys_exit, int status) {
-
-        http_connection_metadata_t *meta = NULL;
-        if (packet_type) {
-            if (packet_type == PACKET_TYPE_RESPONSE) {
-                // if we are filtering by application, ignore the packets not for this connection
-                meta = bpf_map_lookup_elem(&filtered_connections, &conn);
-                //bpf_dbg_printk("Response meta=%lx", meta);
-                if (!meta) {
-                    return 0;
-                }
-            } else { // it must be a request, let's read more bytes
-                u32 full_len = skb->len - tcp.hdr_len;
-                if (full_len > FULL_BUF_SIZE) {
-                    full_len = FULL_BUF_SIZE;
-                }
-                read_skb_bytes(skb, tcp.hdr_len, info->buf, full_len);
-                processing_buf = info->buf;
+        if (size > 0) {
+            void *iovec_ptr = find_msghdr_buf(msg);
+            if (iovec_ptr) {
+                handle_buf_with_connection(&info, iovec_ptr, size, 0);
+            } else {
+                bpf_dbg_printk("can't find iovec ptr in msghdr, not tracking sendmsg");
             }
         }
-        if (packet_type) {
-            bpf_dbg_printk("=== http_filter len=%d pid=%d %s ===", (skb->len - tcp.hdr_len), (meta != NULL) ? meta->pid.host_pid : -1, buf);
-            //dbg_print_http_connection_info(&conn);
+
+        // Checks if it's sandwitched between active SSL handshake uprobe/uretprobe
+        void **s = bpf_map_lookup_elem(&active_ssl_handshakes, &id);
+        if (!s) {
+            return 0;
         }
 
-        process_http(info, &tcp, packet_type, (skb->len - tcp.hdr_len), processing_buf, meta);
+        void *ssl = *s;
+        bpf_dbg_printk("=== kprobe SSL tcp_sendmsg=%d sock=%llx ssl=%llx ===", id, sk, ssl);
+        bpf_map_update_elem(&ssl_to_conn, &ssl, &info, BPF_ANY);
     }
 
     return 0;
 }
 
-// We start by looking when the SSL handshake is established. In between
-// the start and the end of the SSL handshake, we'll see at least one tcp_sendmsg
-// between the parties. Sandwitching this tcp_sendmsg allows us to grab the sock *
-// and match it with our SSL *. The sock * will give us the connection info that is
-// used by the generic HTTP filter.
-SEC("uprobe/libssl.so:SSL_do_handshake")
-int BPF_UPROBE(uprobe_ssl_do_handshake, void *s) {
+//int tcp_recvmsg(struct sock *sk, struct msghdr *msg, size_t len, int flags, int *addr_len)
+SEC("kprobe/tcp_recvmsg")
+int BPF_KPROBE(kprobe_tcp_recvmsg, struct sock *sk, struct msghdr *msg, size_t len, int flags, int *addr_len) {
     u64 id = bpf_get_current_pid_tgid();
 
     if (!valid_pid(id)) {
         return 0;
     }
 
-    bpf_dbg_printk("=== uprobe SSL_do_handshake=%d ssl=%llx===", id, s);
+    bpf_printk("=== tcp_recvmsg id=%d sock=%llx ===", id, sk);
 
-    bpf_map_update_elem(&active_ssl_handshakes, &id, &s, BPF_ANY);
+    // Important: We must work here to remember the iovec pointer, since the msghdr structure
```

Comment:
Saving the msghdr pointer didn't work with requests that are split in few parts, i.e. large requests > 4K in size. Therefore, here I save now the iovec_ptr before we end up receiving the message, so we can read it properly when the first 4K chunk is processed.